### ISSUE TITLE
s3.PresignClient adds an unnecessary extra layer of mocking

### ISSUE THREAD

**Issue by qhenkart** [aws/aws-sdk-go-v2#1081]
Confirm by changing [ ] to [x] below to ensure that it's a bug:
- [ x] I've gone though the [API reference](https://docs.aws.amazon.com/sdk-for-go/v2/api/)
- [ x] I've checked [AWS Forums](https://forums.aws.amazon.com) and [StackOverflow](https://stackoverflow.com/questions/tagged/aws-sdk-go) for answers
- [ x ] I've searched for [previous similar issues](https://github.com/aws/aws-sdk-go-v2/issues) and didn't find any solution
  
**Describe the bug**
I am experimenting with migrating from v1 sdk to v2. The removal of s3iface.API makes sense but as I am going through the functions I ran into https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/s3#NewPresignClient

I don't understand why NewPresignClient is an open function in the s3 package and takes an s3.Client as the first parameter instead of just being a method of the s3.Client. 

This pattern creates a complicated interfacing nightmare to mock these out for tests. The API migration guide suggests making our own interface for s3.Client commands. But since s3.NewPresignClient is not an s3.Client command, and does not accept an interface. Now I will have to create a second interface just for the presignClient and abstract the s3.NewPresignClient in a way that adds unnecessary bloat and complexity to the code

tldr; please make the PresignClient a method of the s3.Client for easy mocking.

**Version of AWS SDK for Go?**
1.0

**Version of Go (`go version`)?**
1.15

**Comment by skmcgrail** [comment]
Our recommendation for mocking the Presign client is that you would define an interface for the Presigner client as well and use that interface in your application so that you can mock out its usage as well.

**Comment by qhenkart** [comment]
@skmcgrail thank you for the response. If the Presign client was uncoupled with the s3.Client, it would be no issue to create a new interface. 

I realize now that the intention is that the presign-client isn't a method that should be called when it is needed but actually in the initial set up/ at the same time as the s3-client is created before it is returned into the interface. I really don't understand why you guys took the direction to create to seemingly identical clients to interact with s3 objects. I'll post my solution here for others who run into this same migration issue

Originally I had a function similar to this 
```
func  NewSession(conf config) (S3Client, error) {
	creds := aws.NewCredentialsCache(credentials.NewStaticCredentialsProvider(conf.Key, conf.Secret,conf.Token 
	_, err := creds.Retrieve(context.TODO())
	if err != nil {
		return nil, errs.ErrInvalidCreds.Context(err)
	}

	cfg, err := config.LoadDefaultConfig(context.Background(), config.WithRegion(conf.Region), config.WithCredentialsProvider(creds))
	if err != nil {
		return nil, errs.ErrInvalidCreds.Context(err)
	}

	return s3.NewFromConfig(cfg), nil
}
```

this returns an S3Client interface. I was hoping to use Presign capabilities from that the s3-client like in the v1 sdk. However in the v2 sdk this is not possible. The Presign client is a separate client that requires the s3-client to initialize. Because of the double interfacing requirement, it would not be clean to create a presign client when needed (without type checks in the production code instead of mocking), it appears necessary to initialize the presign client *before* converting the s3-client into the interface to avoid these type checks. I will use something like this 

```
func  NewSession(conf config) (S3Client, S3PresignClient, error) {
	creds := aws.NewCredentialsCache(credentials.NewStaticCredentialsProvider(conf.Key, conf.Secret,conf.Token 
	_, err := creds.Retrieve(context.TODO())
	if err != nil {
		return nil, errs.ErrInvalidCreds.Context(err)
	}

	cfg, err := config.LoadDefaultConfig(context.Background(), config.WithRegion(conf.Region), config.WithCredentialsProvider(creds))
	if err != nil {
		return nil, errs.ErrInvalidCreds.Context(err)
	}

	return s3.NewFromConfig(cfg), s3.NewPresignClient(s3c), nil
}
```
where S3PresignClient is an interface with the method I need to use

**Comment by jasdel** [comment]
Thanks for the additional information. The presigner is intentionally coupled to the API client due to internal behavior of the API client that is needed, e.g. the Client's `invokeOperation` method. This coupling exists because the API client methods return a type relevant to the API operation call, (e.g. `PutObjectOutput`). Whereas the Presign client does not return the types, and instead returns the [v4.PresignedHTTPRequest](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/aws/signer/v4#PresignedHTTPRequest) type. The `PresignedHTTPRequest` provides the members for presigned URL, and signed headers.

If the presigner were to take an interface type instead of the concrete API client, that interface would need to be made up of a `InvokeOperation` method, and not any specific API operation method.

For these reasons the SDK's `PresignClient` intentionally depends on the API Client due to internal couple that is needed with the API client's internal behavior needed for presigning.

**Comment by qhenkart** [comment]
I have received the guidance and I appreciate that you provided the technical reasons for the decision. I am closing the issue
